Code/GraphMol/Resonance.cpp

```diff
@@ -0,0 +1,1355 @@
+  void AtomElectrons::initTvNbFcFromAtom()
+  {
+    d_tv = d_atom->getTotalValence();
+    d_fc = d_atom->getFormalCharge();
+    d_nb = oe() - d_tv - d_fc;
+  }
+  
+  boost::uint8_t AtomElectrons::findAllowedBonds(unsigned int bi)
+  {
+    // AtomElectrons::findAllowedBonds returns a 6-bit result
+    // encoded as follows:
+    // +-------------------------------------+
+    // | BIT | MEANING                       |
+    // |   1 | can accept single bond        |
+    // |   2 | needs charge if single bonded |
+    // |   3 | can accept double bond        |
+    // |   4 | needs charge if double bonded |
+    // |   5 | can accept triple bond        |
+    // |   6 | needs charge if triple bonded |
+    // +-------------------------------------+
+    
+    allConjBondsDefinitiveBut(bi);
+    boost::uint8_t res = 0;
+    for (unsigned int i = 0; i < 3; ++i) {
+      res |= (canAddBondWithOrder(bi, i + 1) << (i * 2));
+    }
+    return res;
+  }
+
+  // returns true if any conjugated neighbor is charged (and
+  // has atomic number == atomicNum, if the latter is non-zero)
+  bool AtomElectrons::isNbrCharged(unsigned int bo,
+    unsigned int atomicNum)
+  {
+    bool res = false;
+    const ROMol &mol = d_parent->parent()->mol();
+    ROMol::OEDGE_ITER nbrIdx, endNbrs;
+    boost::tie(nbrIdx, endNbrs) = mol.getAtomBonds(d_atom);
+    for (; !res && (nbrIdx != endNbrs); ++nbrIdx) {
+      const Bond *bondNbr = mol[*nbrIdx].get();
+      unsigned int biNbr = bondNbr->getIdx();
+      if (d_parent->parent()->getBondConjGrpIdx(biNbr) != conjGrpIdx())
+        continue;
+      BondElectrons *beNbr =
+        d_parent->getBondElectronsWithIdx(biNbr);
+      const Atom *atomNbr = ((bondNbr->getBeginAtom() == d_atom)
+        ? bondNbr->getEndAtom() : bondNbr->getBeginAtom());
+      unsigned int aiNbr = atomNbr->getIdx();
+      AtomElectrons *aeNbr =
+        d_parent->getAtomElectronsWithIdx(aiNbr);
+      res = (((beNbr->isDefinitive() && !aeNbr->hasOctet())
+        || (!beNbr->isDefinitive() && aeNbr->isDefinitive()
+        && (aeNbr->oe() < (5 - bo))))
+        && (!atomicNum || (atomNbr->getAtomicNum() == atomicNum)));
+    }
+    return res;
+  }
+  
+  // returns a 2-bit value, where the least significant bit is true
+  // if the atom can add a bond with order bo, and the most significant
+  // bit is true if the atom needs be charged
+  boost::uint8_t AtomElectrons::canAddBondWithOrder(unsigned int bi,
+    unsigned int bo)
+  {
+    boost::uint8_t canAdd = !isDefinitive();
+    if (canAdd)
+      canAdd = (d_tv <= (5 - bo));
+    // if canAdd is true, loop over neighboring conjugated bonds
+    // and check their definitive flag; if all neighbors are
+    // definitive, we need an additional check on total valence
+    if (canAdd && isLastBond()) {
+      // we allow a formal charge up to 2 on atoms right of
+      // carbon, not more than 1 on atoms left of N
+      unsigned int rightOfC = ((d_atom->getAtomicNum() > 6) ? 1 : 0);
```

Comment:
I haven't found a counter-example yet, but us it this going to work for atoms not in the first row of the periodic table like Si?